packages/core/contracts/financial-templates/demo/OptimisticDepositBox.sol

```diff
@@ -0,0 +1,290 @@
+    event RequestWithdrawalExecuted(
+        address indexed user,
+        uint256 indexed collateralAmount,
+        uint256 exchangeRate,
+        uint256 withdrawalRequestTimestamp
+    );
+    event RequestWithdrawalCanceled(
+        address indexed user,
+        uint256 indexed collateralAmount,
+        uint256 withdrawalRequestTimestamp
+    );
+
+    /****************************************
+     *               MODIFIERS              *
+     ****************************************/
+
+    modifier noPendingWithdrawal(address user) {
+        _depositBoxHasNoPendingWithdrawal(user);
+        _;
+    }
+
+    /****************************************
+     *           PUBLIC FUNCTIONS           *
+     ****************************************/
+
+    /**
+     * @notice Construct the OptimisticDepositBox.
+     * @param _collateralAddress ERC20 token to be deposited.
+     * @param _finderAddress UMA protocol Finder used to discover other protocol contracts.
+     * @param _priceIdentifier registered in the DVM, used to price the ERC20 deposited.
+     * The price identifier consists of a "base" asset and a "quote" asset. The "base" asset corresponds
+     * to the collateral ERC20 currency deposited into this account, and it is denominated in the "quote"
+     * asset on withdrawals.
+     * An example price identifier would be "ETH/USD" which will resolve and return the USD price of ETH.
+     * @param _timerAddress contract that stores the current time in a testing environment.
+     * Must be set to 0x0 for production environments that use live time.
+     */
+    constructor(
+        address _collateralAddress,
+        address _finderAddress,
+        bytes32 _priceIdentifier,
+        address _timerAddress
+    ) nonReentrant() Testable(_timerAddress) {
+        collateralCurrency = IERC20(_collateralAddress);
+        priceIdentifier = _priceIdentifier;
+        finder = FinderInterface(_finderAddress);
+    }
+
+    /**
+     * @notice Transfers `collateralAmount` of `collateralCurrency` into caller's deposit box.
+     * @dev This contract must be approved to spend at least `collateralAmount` of `collateralCurrency`.
+     * @param collateralAmount total amount of collateral tokens to be sent to the sponsor's position.
+     */
+    function deposit(uint256 collateralAmount) public nonReentrant() {
+        require(collateralAmount > 0, "Invalid collateral amount");
+        OptimisticDepositBoxData storage depositBoxData = depositBoxes[msg.sender];
+        if (depositBoxData.collateral == 0) {
+            emit NewOptimisticDepositBox(msg.sender);
+        }
+
+        // Increase the individual deposit box and global collateral balance by collateral amount.
+        depositBoxData.collateral = depositBoxData.collateral.add(collateralAmount);
+        require(depositBoxData.collateral > 0, "Collateral not added");
```

Comment:
Is it possible for this to fail?

Reply:
Hmm, I don't think so. I'll remove the check.